File: Java/Migration/unit-tests-java/src/test/java/p04_mocks/Ex04NoMoreInteractions.java
package p04_mocks;

import org.testng.annotations.Test;
import p04_mockito.CommandInterpreter;
import p04_mockito.ExecutionEngine;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;

public class Ex04NoMoreInteractions {

  @Test
  public void shouldTranslateShootingCommandToShootingReport() {
    //GIVEN
    ExecutionEngine executionEngine = mock(ExecutionEngine.class);
    CommandInterpreter interpreter = new CommandInterpreter(executionEngine);

    //WHEN
    interpreter.onEvent(CommandInterpreter.SHOOTING);

    //THEN
    verify(executionEngine).shootingStarted();
    verifyNoMoreInteractions(executionEngine);
  }

  @Test
  public void shouldTranslateRobberyCommandToRobberyReport() {
    //GIVEN
    ExecutionEngine executionEngine = mock(ExecutionEngine.class);
    CommandInterpreter interpreter = new CommandInterpreter(executionEngine);

    //WHEN
    interpreter.onEvent(CommandInterpreter.ROBBERY);

    //THEN
    verify(executionEngine).robberyTookPlace();
    verifyNoMoreInteractions(executionEngine);
  }

  @Test
  public void shouldTranslateIncidentCommandToIncidentReport() {
    //GIVEN
    ExecutionEngine executionEngine = mock(ExecutionEngine.class);
    CommandInterpreter interpreter = new CommandInterpreter(executionEngine);

    //WHEN
    interpreter.onEvent(CommandInterpreter.INCIDENT);

    //THEN
    verify(executionEngine).incidentDiscovered();
    verifyNoMoreInteractions(executionEngine /*, executionEngine, executionEngine */ );
  }
}

File: Java/Migration/unit-tests-csharp/unit-tests-csharp/P04NSubstitute/Ex04NoMoreInteractions.cs
using NSubstitute;
using NUnit.Framework;
using TddEbook.TddToolkit.NSubstitute;
using unit_tests_csharp.P04NSubstitute.Ex04ProductionCode;

namespace unit_tests_csharp.P04NSubstitute
{
  public class Ex04NoMoreInteractions
  {

    [Test]
    public void ShouldTranslateShootingCommandToShootingReport()
    {
      //GIVEN
      var executionEngine = Substitute.For<IExecutionEngine>();
      var interpreter = new CommandInterpreter(executionEngine);

      //WHEN
      interpreter.OnEvent(CommandInterpreter.Shooting);

      //THEN
      XReceived.Only(() =>
      {
        executionEngine.ShootingStarted();
      });
    }

    [Test]
    public void ShouldTranslateRobberyCommandToRobberyReport()
    {
      //GIVEN
      var executionEngine = Substitute.For<IExecutionEngine>();
      var interpreter = new CommandInterpreter(executionEngine);

      //WHEN
      interpreter.OnEvent(CommandInterpreter.Robbery);

      //THEN
      XReceived.Only(() =>
      {
        executionEngine.RobberyTookPlace();
      });
    }

    [Test]
    public void ShouldTranslateIncidentCommandToIncidentReport()
    {
      //GIVEN
      var executionEngine = Substitute.For<IExecutionEngine>();
      var interpreter = new CommandInterpreter(executionEngine);

      //WHEN
      interpreter.OnEvent(CommandInterpreter.Incident);

      //THEN
      XReceived.Only(() =>
      {
        executionEngine.IncidentDiscovered();
      });
    }
  }
}